File: src/main/java/dev/akif/crud/error/CRUDErrorException.java
package dev.akif.crud.error;

import lombok.EqualsAndHashCode;
import lombok.Value;
import org.springframework.http.HttpStatus;

/**
 * Wrapper for a {@link CRUDError} as a {@link RuntimeException} so it can be thrown
 * <p>
 * The stack trace and suppression is disabled by default.
 */
@EqualsAndHashCode(callSuper = true)
@Value
public class CRUDErrorException extends RuntimeException {
    /**
     * {@link CRUDError} that occurred
     */
    CRUDError error;

    /**
     * Constructor to create a {@link CRUDErrorException} from given error
     *
     * @param error {@link CRUDError}
     */
    public CRUDErrorException(final CRUDError error) {
        super(error.toString(), null, false, false);
        this.error = error;
    }

    /**
     * Builds an "already exists" error as a {@link CRUDErrorException} with "409 Conflict" HTTP status for given data
     *
     * @param what Type of what already exists
     * @param data Data describing what already exists
     * @param <D>  Type of the data describing what already exists
     * @return {@link CRUDErrorException} containing {@link CRUDError} built from given data
     */
    public static <D> CRUDErrorException alreadyExists(final String what, final D data) {
        return new CRUDErrorException(
                new CRUDError(HttpStatus.CONFLICT, "%s with %s already exists.".formatted(what, data))
        );
    }

    /**
     * Builds a "not found" error as a {@link CRUDErrorException} with "404 Not Found" HTTP status for given id
     *
     * @param what Type of what is not found
     * @param id   Id of what is not found
     * @param <I>  Type of the id of what is not found
     * @return {@link CRUDErrorException} containing {@link CRUDError} built from given data
     */
    public static <I> CRUDErrorException notFound(final String what, final I id) {
        return new CRUDErrorException(
                new CRUDError(HttpStatus.NOT_FOUND, "%s with id %s is not found.".formatted(what, id))
        );
    }
}

File: api/src/main/kotlin/dev/akif/crud/error/CRUDErrorException.kt
package dev.akif.crud.error

import org.springframework.http.HttpStatus

/**
 * Wrapper for a [CRUDError] as a [RuntimeException] so it can be thrown
 *
 * The stack trace and suppression is disabled by default.
 *
 * @property error [CRUDError] to wrap
 */
data class CRUDErrorException(val error: CRUDError) : RuntimeException(error.toString(), null, false, false) {
    /** @suppress */
    companion object {
        /**
         * Builds an "already exists" error as a [CRUDErrorException] with "409 Conflict" HTTP status for given data
         *
         * @param what Type of what already exists
         * @param data Data describing what already exists
         * @return [CRUDErrorException] containing [CRUDError] built from given data
         */
        @JvmStatic
        fun alreadyExists(what: String, data: Any): CRUDErrorException {
            return CRUDErrorException(
                CRUDError(HttpStatus.CONFLICT, "$what with $data already exists.")
            )
        }

        /**
         * Builds a "not found" error as a [CRUDErrorException] with "404 Not Found" HTTP status for given id
         *
         * @param what Type of what is not found
         * @param id   Id of what is not found
         * @return [CRUDErrorException] containing [CRUDError] built from given data
         */
        @JvmStatic
        fun notFound(what: String, id: Any): CRUDErrorException {
            return CRUDErrorException(
                CRUDError(HttpStatus.NOT_FOUND, "$what with id $id is not found.")
            )
        }
    }
}
